src/inet/UDPEndPointImplLwIP.cpp

```diff
@@ -278,6 +283,9 @@ void UDPEndPointImplLwIP::Free()
 
 void UDPEndPointImplLwIP::HandleDataReceived(System::PacketBufferHandle && msg, IPPacketInfo * pktInfo)
 {
+    // Clear mDelayRelease flag so that we don't need to delay release of current UDPEndPoint.
+    mDelayRelease = false;
```

Comment:
Is it possible to get multiple `HandleDataReceived` calls until ep->HandleDataReceived is scheduled by ScheduleLambda?

Wondering what makes it sufficient to have a bool mDelayRelease instead of something like std::atomic<int> ?

Reply:
+1 on concurrency. This is already why we have to defer release

Reply:
Acked

Reply:
I was not only considering concurrency, but also races. Thinking of this sequence:

UDP packet received -> UDPEndPointImplLwIP::LwIPReceiveUDPMessag -> ScheduleLambda (mDelay = true, queue size 1)
UDP packet received -> UDPEndPointImplLwIP::LwIPReceiveUDPMessag -> ScheduleLambda (mDelay = true, queue size 2)
System thread wakes up -> UDPEndPointImplLwIP::HandleDataReceived (mDelay = false, queue size 1!)
UDPEndPointImplLwIP::CloseImpl (mDelay == false) -> Release
System thread wakes up -> UDPEndPointImplLwIP::HandleDataReceived on freed up endpoint

Reply:
I need to understand that we guarantee that `UDPEndPointImplLwIP::LwIPReceiveUDPMessage` cannot be called twice without `UDPEndPointImplLwIP::HandleDataReceived` having been picked up.

Reply:
I guess we can not, since ScheduleLambda scheduled to be executed asynchronously. How about using reference counting std::atomic mDelayReleaseCount?
UDP packet received -> UDPEndPointImplLwIP::LwIPReceiveUDPMessag -> ScheduleLambda (mDelay = 1, queue size 1)
UDP packet received -> UDPEndPointImplLwIP::LwIPReceiveUDPMessag -> ScheduleLambda (mDelay = 2, queue size 2)
System thread wakes up -> UDPEndPointImplLwIP::HandleDataReceived (mDelay - 1, queue size 1!)
UDPEndPointImplLwIP::CloseImpl (mDelay == 1) 
System thread wakes up -> UDPEndPointImplLwIP::HandleDataReceived (mDelay - 1, queue size 1!)
UDPEndPointImplLwIP::CloseImpl (mDelay == 0) -> Release

Reply:
This is what I was proposing. My original comment had `std::atomic<int>` in the text but without backticks it seems the markdown handler just swallows everything between `<>`